[Update #26] Add HtmlHelper test cases

gundamew opened the pull request:
For #26
1. Add HtmlHelper::repair test cases.
2. Add HtmlHelper::getJSObject test case.

Commit:
Add HtmlHelper test cases

1. Add HtmlHelper::repair test cases.
2. Add HtmlHelper::getJSObject test case.

LeoOnTheEarth commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,203 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+	 * @covers       Windwalker\Helper\HtmlHelper::repair
 	 */
-	public function testRepair()
+	public function testRepairHtmlClosedTidy($html_string_closed)
```
變數命名要用 camel 原則

ex: `fooBarBalabala`

LeoOnTheEarth replied:
method 參數應該還要帶一個 `$expected` 進來 (參考一下 ArrayHelperTest 吧)

先看一下這個 manual: http://php.net/manual/en/tidy.repairstring.php 

`tidy_repair_string` 的目的是去修正不正確的 close tag
你應該要帶兩個參數到 `testRepairHtmlClosedTidy` 內
- 預期被 `HtmlHelper::repair` 改過的字串
- 要被測試的原始字串，這要有錯誤的 close tag 以確定 `HtmlHelper::repair` 有正確的修補過

LeoOnTheEarth replied:
看一下 manual 的 example 會比較知道要怎麼設計測試的 case 喔

LeoOnTheEarth replied:
哈，看到下面有測試到沒被 close 的 tag
那就沒事~

gundamew replied:
變數命名原則在 [smstw/sms-coding-standard](https://github.com/smstw/sms-coding-standard) 裡沒有被提到，所以我參考了其他程式碼裡的變數命名方式

LeoOnTheEarth replied:
確實是沒被提到，不過我們這邊是習慣用 camel 的命名方式
我晚點丟一下 coding standard 的修改好了
目前只提到了再 Joomla CMS 下的變數命名原則
應該要再補上 SMS Taiwan 這邊在用的

gundamew replied:
了解 :+1:

asika32764 replied:
嚴格說起來我們沒有規定變數命名

但我們有規範類別屬性在專案中統一用一種模式，所以變數也是
https://github.com/smstw/sms-coding-standard#%E9%A1%9E%E5%88%A5%E5%B1%AC%E6%80%A7

既然多數變數都是馱峰，則自然以馱峰式為準

LeoOnTheEarth replied:
加上變數命名原則了

https://github.com/smstw/sms-coding-standard/pull/10

gundamew replied:
:+1:

LeoOnTheEarth commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,203 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+	 */
+	public function testRepairHtmlUnclosed($html_string_unclosed)
+	{
+		$this->assertNotSame($html_string_unclosed, HtmlHelper::repair($html_string_unclosed, false));
```
這個測試不會過喔

原因的解釋請看這邊 https://github.com/smstw/windwalker-joomla-rad/pull/34/files#r27275924

LeoOnTheEarth commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,203 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+  Over my dead body
+</p>
+HTML_STR_CLOSED_1
+),
```
縮排要注意一下

Commit:
Refactor test cases

1. Changing indentation.
2. Refactor data providers.
    1. Change test case arguments.
    2. Add HtmlHelperTest::repairHtmlUnclosedTidyDataProvider.

Commit:
Refactor data providers (nested array).

LeoOnTheEarth commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,292 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+	 *
+	 * @return array $returns
+	 */
+	public function repairHtmlClosedDataProvider()
```
要不要加一個排版比較亂的測試?

類似下面這樣

``` html
    <p>
Over my dead body
</p>
```

gundamew replied:
@LeoOnTheEarth 

試著測試了一下，Tidy 的部份沒問題，但是不用 Tidy 的話就無法處理了：

``` bash
# 測試案例
      <table>
  <tr><td>
    Over my dead body</td>
    </tr>
 </table>

# 測試結果
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-<table>
-  <tr>
-    <td>
-      Over my dead body
-    </td>
-  </tr>
-</table>
+      <table>
+  <tr><td>
+    Over my dead body</td>
+    </tr>
+ </table>
```

LeoOnTheEarth replied:
沒用 tidy 的部分可以先把測試擺著
程式碼的部分看要怎麼去修正
或是沒必要修正的話，就先用目前可預期的輸出做測試吧

gundamew replied:
:ok_hand: 已改

asika32764 commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,292 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+	public function testGetJSObject($array)
+	{
+		json_decode(HtmlHelper::getJSObject($array));
+		$this->assertTrue(json_last_error() === JSON_ERROR_NONE);
```
其實可以不需要這個test，只要 parse error 一定不會有任何值出現，所以只要判斷decode的值正不正確，就可以完全 cover 這個 test method

例如

``` php
testGetJSObject($array, $result)
{
    $this->assertEquals($result, json_decode(HtmlHelper::getJSObject($array), true));
}
```

LeoOnTheEarth replied:
我是想說應該針對 `HtmlHelper::getJSObject($array)` 的 return 值做比對會比較完整
只檢查 json_decode 只能確保 JSON 格式正確，但不能確保 `HtmlHelper::getJSObject($array)` 回傳的值是符合預期的

asika32764 replied:
是的，所以只要 return 值比對沒錯，表示 json 格式一定正確，所以無需多判斷 parse error 的問題。

gundamew replied:
我直接在 expected 裡存 JSON 字串了 :stuck_out_tongue:

asika32764 commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,292 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+	 */
+	public function testRepairHtmlUnclosedTidy($expected, $data)
+	{
+		$this->assertSame($expected, HtmlHelper::repair($data));
```
由於 tidy 並非在所有 php 環境都有預設安裝，建議在所有 tidy test 中增加

``` php
if (!function_exists('tidy_repair_string'))
{
    $this->markThisTestSkip('Tidy not exists');

    return;
}
```

gundamew replied:
已修改

Commit:
[Update #22] More modifications

1. Add new test data for closed HTML tags.
2. Add data provider for closed HTML tags test case w/o Tidy.
3. Add statement for checking if Tidy was available.
4. Refactoring HtmlHelper::getJSObject test case.

LeoOnTheEarth commented:
:+1:

asika32764 commented on test/Helper/HtmlHelperTest.php:
```diff
@@ -24,36 +26,396 @@ class HtmlHelperTest extends \PHPUnit_Framework_TestCase
+		else
+		{
+			$this->assertSame($expected, HtmlHelper::repair($data));
+		}
```
這種狀況不需要else

Return 就好

LeoOnTheEarth replied:
是要在 if 裡面 return 吧?

asika32764 replied:
Yes

Leo notifications@github.com 於 2015年3月31日 星期二寫道：

> In test/Helper/HtmlHelperTest.php
> https://github.com/smstw/windwalker-joomla-rad/pull/34#discussion_r27470455
> :
> 
> > ```
> > {
> > ```
> > -       // Remove the following lines when you implement this test.
> > -       $this->markTestIncomplete(
> > -           'This test has not been implemented yet.'
> > -       );
> > -       if (!function_exists('tidy_repair_string'))
> > -       {
> > -           $this->markTestSkipped(
> > -               'The Tidy extension is not available.'
> > -           );
> > -       }
> > -       else
> > -       {
> > -           $this->assertSame($expected, HtmlHelper::repair($data));
> > -       }
> 
> 是要在 if 裡面 return 吧?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/smstw/windwalker-joomla-rad/pull/34/files#r27470455.

## 

Best Regard

## Simon Asika | Kwang Po Tung | 董冠伯
- Founder: Asikart Extension http://ext.asikart.com/
- CTO: SMS Taiwan http://www.bm-sms.com.tw/
- Mobile: 886-928-565-747
- Github: https://github.com/asika32764

gundamew replied:
另外發了 PR #38 處理這個修改
